### ISSUE TITLE
clang reports warnings

### ISSUE THREAD

**Issue by bmorel** [ndilieto/uacme#37]
Hello.
Just found your client, and decided that before using it, it would be interesting to see what clang --analyze would say about it.

Turns out that there are several warnings here (I will join the full log and how to generate it, since clang joins the traces it took to find the warnings):

* null pointer passed as argument to strcasecmp/strdup
* Potential leak of memory
* Attempt to free released memory
* The left operand of '==' is a garbage value
* Access to field 'tls' results in a dereference of a null pointer

Here is how I generated that:
`clang -D SYSCONFDIR='"/etc/"' -D RUNSTATEDIR='"/var/run"' -I./libev --analyze --analyzer-output text *.c > build.log 2>&1`

I have also noticed warnings with a different combination of flags:

* reports expression results being unused in the asserts. Nice trick you have, but could I suggest using `assert( condition && "error message" );` instead? It does the same, but I still have to see a compiler whining about it.
* implicit conversion changes signedness
* macro name is a reserved identifier
* format string is not a string literal
* implicit conversion loses integer precision:
* enumeration values not explicitly handled in switch (but the default keyword is used at least sometimes, so...)
* no previous prototype for function
* declaration shadows a local variable
* cast from 'struct sockaddr *' to 'struct sockaddr_in *' increases required alignment (can't really be helped though)
* padding size of 'curldata_t' with 4 bytes to alignment boundary (again, probably can't be helped, and this one only impacts memory impact AFAIK)
* disabled expansion of recursive macro (I still have to understand the impact of this one)
* 'HAVE_MAP_DEVZERO' is not defined, evaluates to 0 (this is my fault, since I short-circuited the makefile)
* cast from 'const void *' to 'unsigned char *' drops const qualifier
* cast from 'uint8_t *' (aka 'unsigned char *') to 'client_t *' (aka 'struct client *') increases required alignment from (this might be ok, but it also might trigger bugs, and that one could maybe be avoided)
* cast from function call of type 'long' to non-matching type 'float'
* implicit conversion increases floating-point precision:

Of course, several of those warnings have only memory performance impact or are false positives, but I believe fixing most of those would allow an easier to maintain code (shadow declarations, names reserved by the standard,...) or less prone to errors (implicit casts, for example).

Those were generated with:
`clang -Wall -Weverything -D SYSCONFDIR='"/etc/"' -D RUNSTATEDIR='"/var/run"' -I./libev  *.c > build2.log 2>&1`
Of course, this generates linking error too, but that's unimportant considering my goal here.

You will find both the analyzer log (build.log) and the clang complains (build2.log) attached.
I intend to check if I can patch some of those quickly.

[build.tar.gz](https://github.com/ndilieto/uacme/files/4713879/build.tar.gz)

**Comment by bmorel** [comment]
I took some time to reduce the number of warnings emitted by clang --analyzer, I am joining the patch to this comment.

To be honest, it's mostly marking where a patch would actually solve the issue. For several of those, I have no idea how to fix properly the problem.

There are still 2 potential double free and one garbage comparison that I could not find where to patch to avoid them.

Finally, I fixed a false positive using a C11 feature: `_Noreturn`. I do not know if you support it though.
[some.analysis.patch.tar.gz](https://github.com/ndilieto/uacme/files/4714206/some.analysis.patch.tar.gz)

**Comment by ndilieto** [comment]
TL;DR waste of my time

Please note that I check the code for memory leaks, uninitialized variable assignments and other unpleasantries using [valgrind](https://valgrind.org), which unlike the analyzer you used actually runs the code instead of trying to figure it out by just looking at it.

I've checked the supposed "problems" you reported and they are all false alarms. All are notices, only three are warnings. As you have wasted enough of my scarce time and haven't even bothered to understand what the code does, I will only comment on the first warning for the sake of transparency. Here is the analysis:

`crypto.c:2396:25: warning: Null pointer passed as an argument to a 'nonnull' parameter`

The analyzer is ignoring context and can't figure out that the strcasecmp call it laments about  happens only when when the earlier calls to gnutls_x509_crq_get_dn_by_oid or gnutls_x509_crq_get_subject_alt_name have returned a success code, which means buf is guaranteed not to be null to start with. The falltthrough from GNUTLS_SAN_IPADDRESS, which you'll see I marked as intentional, happens when buf is first freed then replaced by ip, which also is checked for null beforehand.

```
    do {
        r = cn ?
            gnutls_x509_crq_get_dn_by_oid(crq, GNUTLS_OID_X520_COMMON_NAME,
                    ncn, 0, buf, &size) :
            gnutls_x509_crq_get_subject_alt_name(crq, nsan, buf, &size,
                    NULL, NULL);

        switch (r) {
            case GNUTLS_E_REQUESTED_DATA_NOT_AVAILABLE:
                if (cn) {
                    cn = false;
                    r = GNUTLS_E_SUCCESS;
                }
                break;

            case GNUTLS_E_SHORT_MEMORY_BUFFER:
                buf = calloc(1, size);
                if (!buf) {
                    warn("csr_names: calloc failed");
                    goto out;
                }
                break;

            case GNUTLS_SAN_IPADDRESS:
                ip = calloc(1, INET6_ADDRSTRLEN);
                if (!ip) {
                    warnx("csr_names: calloc failed");
                    goto out;
                }
                if (!inet_ntop(size == 4 ? AF_INET : AF_INET6, buf, ip,
                            INET6_ADDRSTRLEN)) {
                    warnx("csr_names: invalid IP address in Subj Alt Name");
                    free(ip);
                    ip = NULL;
                    continue;
                }
                free(buf);
                buf = ip;
                ip = NULL;
                // intentional fallthrough
            case GNUTLS_E_SUCCESS:
            case GNUTLS_SAN_DNSNAME:
                for (i = 0; i < n; i++) {
                    if (strcasecmp(buf, names[i]) == 0)
                        break;
                }
```
